### PR TITLE
gigaspeech

### PR DESCRIPTION
**Pull request by sw005320**
- [x] add a result
- [x] add the gigaspeech variable to `db.sh`
- [x] add a corpus description to https://github.com/espnet/espnet/blob/master/egs2/README.md

### CODE REVIEW

File: egs2/gigaspeech/asr1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    [ ! -f ${meta_dir}/utt2subsets ] && echo "Error: No such file ${meta_dir}/utt2subsets!" && exit 1
+    for sub in ${!dic_sets[*]}; do
+	[ ! -d ${data_dir}/${sub} ] && mkdir -p ${data_dir}/${sub}
+	tag=${dic_sets[${sub}]}
+	grep "{$tag}" ${meta_dir}/utt2subsets | subset_data_dir.sh --utt-list - ${data_dir}/corpus ${data_dir}/${sub}
+	utils/fix_data_dir.sh ${data_dir}/${sub}
+    done
+fi
+
+log "Successfully finished. [elapsed=${SECONDS}s]"
```
Comment (chenguoguo) [inline egs2/gigaspeech/asr1/local/data.sh:92]: I'd suggest we keep data preparation related scripts to GigaSpeech's repo, e.g., toolkits/espnet/gigaspeech_data_prep.sh, and in egs2/gigaspeech/asr1/local/data.sh, we simply clone the GigaSpeech repo, and call the script directly. Of course, you could add version check to make sure that you are getting the expected version of the data.

The reason is that we may update GigaSpeech over the time, and if that happens, we only have to update scripts in GigaSpeech.

**Review by chenguoguo — COMMENTED** [review]
Generally looks good to me. Added one comment.

File: egs2/gigaspeech/asr1/local/data.sh (hunk 1 above)
Comment (sw005320) [reply egs2/gigaspeech/asr1/local/data.sh:92]: Definitely.
The reason I copied was that the data preparation script was not in a good shape.
Now, https://github.com/SpeechColab/GigaSpeech/blob/main/toolkits/kaldi/gigaspeech_data_prep.sh looks good and I think I can manage to just call it.

Comment (sw005320) [reply egs2/gigaspeech/asr1/local/data.sh:69]: How about the latest change?
(although I want you to fix https://github.com/SpeechColab/GigaSpeech/issues/16)

Comment (chenguoguo) [reply egs2/gigaspeech/asr1/local/data.sh:69]: Looks good to me! We'll fix that issue on GigaSpeech

**Comment by dophist** [comment]
Just tested GigaSpeech's Podcast DevSet(GigaSpeech.json v0.2.2) with Google Cloud Speech API, got 12% WER.

After digging into WER log, I found a remarkable portion of errors came from words like "__AH__, __UM__, __UH__" and disfluencies like "__It's__ __it's__ it's a ...", Google must have adopted a text post-processing module towards raw ASR hyp. 

From recipe/pipeline perspective, we could incorporate a similar post-processing module into scoring stage(including certain types of TN, disfluency smoothing, punctuations, capitalization etc). This will make resulting WER more stable and robust in spontaneous ASR scenarios.

We may start with some naive implementations of post-processing module in GigaSpeech repo, which can be applied to REF & HYP text before WER scoring, any ideas and suggestions on this, let us know. @sw005320

**Comment by dophist** [comment]
@sw005320 By the way, we intend to borrow librispeech's test sets as GigaSpeech's Audio-Book-Domain test sets, it would be great if you can test test-clean & test-other on already trained GigaSpeech model when you have time.

**Comment by sw005320** [comment]
This is the result.
I think it is reasonable as a result without LM.

## conformer, bpe5000, epoch 16, no LM
### WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|gigaspeech_dev|2043|51075|92.8|4.6|2.6|2.2|9.3|66.4|
|gigaspeech_test|9627|175116|90.4|7.1|2.5|6.3|15.9|69.7|
|librispeech_dev_clean|2703|54402|95.5|3.0|1.4|0.2|4.7|39.0|
|librispeech_dev_other|2864|50948|92.4|6.2|1.4|0.6|8.3|55.2|
|librispeech_test_clean|2620|52576|95.8|2.9|1.3|0.3|4.4|37.2|
|librispeech_test_other|2939|52343|92.5|6.1|1.5|0.7|8.2|56.5|

**Comment by sw005320** [comment]
I also post the current issue (I already sent the same content by email). Just to record it in this repo.

I found three typical issues in the reference (see below).
Since e2e ASR is really good, you can actually find some reference issues by checking the bad ASR result utterances (ASR can detect the reference error).
I’m attaching the ASR result, so I’d suggest double-checking all references by using this result.

Example 1)
I found that most of the error comes from the inconsistent hyphen and so on, and we could remove it during the scoring.
I want someone to consider implementing it during the scoring.

```
id: (pod1000000005_s0000428-pod1000000005_s0000428)
Scores: (#C #S #D #I) 11 2 0 2
REF: after ****** THIRTY-FIVE years control of forever ****** TWENTY-ONE slipped through the chang family's hands
HYP: after THIRTY FIVE years control of forever TWENTY ONE slipped through the chang family's hands
```

Example 2)
The most frequent errors.
No words in the reference. I’m not sure, but it may happen in the data preparation or the original transcription.
I don’t have time to check it
I picked up very storage cases, but basically, you should double-check the part that has very large insertion errors.

```
Speaker sentences1969: pod1000000016_s0000110 #utts: 1
id: (pod1000000016_s0000110-pod1000000016_s0000110)
Scores: (#C #S #D #I) 0 0 0 20
REF: ******* ********** ****** **** ***** ***** **** ***** ***** *** **** **** ******* ******* **** *** *** ******* ** *********
HYP: CERTAIN UNALLIABLE RIGHTS THAT AMONG THESE THAT AMONG THESE ARE LIFE LIFE LIBERTY LIBERTY SUIT AND THE PURSUIT OF HAPPINESS
Eval: I I I I I I I I I I I I I I I I I I I I

Speaker sentences2000: pod1000000016_s0000144 #utts: 1
id: (pod1000000016_s0000144-pod1000000016_s0000144)
Scores: (#C #S #D #I) 0 0 0 66
REF: ** *** ********* *** *** * ****** ** ***** **** ** **** **** * *** *** **** ** **** ****** ***** ** **** ** ** * ****** ***** * **** ***** ******* *** **** ** *** *** **** ***** **** **** *** ***** ** ** ************** ** ** * ***** *** **** ** *** *** *** *** **** *** *********** *** *** * *********** *** *******************
HYP: UM AND SOMETIMES YOU SAY I WONDER IF WE'LL EVER BE FREE WHEN I ASK THE LORD TO FREE YOU'RE GOING TO SING IT OH I WONDER SHALL I EVER REACH HEAVENS I'M AWED IF I'M AND THEY WOULD SING THAT FOR ABOUT AN OW WEEHHHHHHHHHHH IT OR I CAN'T I'T SING IT BUT I'T I'M AWS LIKE I'M AWOOHHHHHHH I'M NOT A HOOHHHHHHHH I'M AWHHHHHHHHHHHHHHHHH
Eval: I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I

Speaker sentences2062: pod1000000016_s0000227 #utts: 1
id: (pod1000000016_s0000227-pod1000000016_s0000227)
Scores: (#C #S #D #I) 0 0 0 74
REF: *** *** ***** **** ***** **** ** *** ***** **** *** ** **** ******** *** ***** ******* **** **** *********** ******** ** ** ******** **** *** ** **** ****** ******** *** **** **** ** **** * ***** **** **** *** ***** ** ***** ** ******* *** ***** **** ** *** ***** ** ******** ********** ******* **** ** ****** **** * ***** ******* ***** ** ***** ********* *** ******** ****** *** ** **** *** *****
HYP: YOU CAN NEVER WHIP THESE BUGS IF YOU DON'T KEEP YOU IN THEM SEPARATE THE WHOLE TROUBLE WITH THIS INTEGRATION BUSINESS IS IT PROBABLY WILL END UP WITH MIXING SOCIALLY BUT THEY TELL ME THAT I DON'T EVEN HAVE THE RIGHT TO FIGHT TO PROTECT THE WHITE RACE WE ARE GOING TO MAINTAIN SEGREGATED SCHOOLS DOWN IN DEXSEI WELL I THINK THEY'RE AIMED AS MIXED MARRIAGES AND BECOMING YOU'VE GOT TO KEEP THE WHOLE
Eval: I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I
 I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I I
```

Example 3)
The entire deletion.
Probably, in this case, the audio was too short.

```
Speaker sentences3741: pod1000000026_s0000036 #utts: 1
id: (pod1000000026_s0000036-pod1000000026_s0000036)
Scores: (#C #S #D #I) 0 1 5 0
REF: THAT'S WHAT I SAID VERY BASIC
HYP: ****** **** * **** **** <SOS/EOS>
Eval: D D D D D S
```

**Comment by sw005320** [comment]
> We may start with some naive implementations of post-processing module in GigaSpeech repo, which can be applied to REF & HYP text before WER scoring, any ideas and suggestions on this, let us know. @sw005320

Actually, my suggestion is to minimize such post-processing.
As long as we use the same post-processing, we can compare the result.
We could consider hyphen and apostrophe like the other corpus is doing.
We may also normalize the backchannel. We could use the rule in the Kaldi switchboard recipe (https://github.com/kaldi-asr/kaldi/blob/master/egs/swbd/s5c/local/score_sclite.sh#L66-L84).
But basically, that's it.

**Comment by chenguoguo** [comment]
I added this issue to GigaSpeech: https://github.com/SpeechColab/GigaSpeech/issues/24

I think it makes sense to have GigaSpeech to provide the normalization/scoring scripts so that results will be comparable across different toolkits.

**Comment by dophist** [comment]
> Example 1)
> I found that most of the error comes from the inconsistent hyphen and so on, and we could remove it during the scoring.
> I want someone to consider implementing it during the scoring.
> 
> ```
> id: (pod1000000005_s0000428-pod1000000005_s0000428)
> Scores: (#C #S #D #I) 11 2 0 2
> REF: after ****** THIRTY-FIVE years control of forever ****** TWENTY-ONE slipped through the chang family's hands
> HYP: after THIRTY FIVE years control of forever TWENTY ONE slipped through the chang family's hands
> ```

Yes exactly, we can have hyphen removed in post-processing, will do.

> Example 2)
> The most frequent errors.
> No words in the reference. I’m not sure, but it may happen in the data preparation or the original transcription.
> I don’t have time to check it
> I picked up very storage cases, but basically, you should double-check the part that has very large insertion errors.
> 
> ```
> Speaker sentences1969: pod1000000016_s0000110 #utts: 1
> id: (pod1000000016_s0000110-pod1000000016_s0000110)
> Scores: (#C #S #D #I) 0 0 0 20
> REF: ******* ********** ****** **** ***** ***** **** ***** ***** *** **** **** ******* ******* **** *** *** ******* ** *********
> HYP: CERTAIN UNALLIABLE RIGHTS THAT AMONG THESE THAT AMONG THESE ARE LIFE LIFE LIBERTY LIBERTY SUIT AND THE PURSUIT OF HAPPINESS
> Eval: I I I I I I I I I I I I I I I I I I I I


We checked this, this is due to that we have "invalid speech utterances" that are annotated with "garbage tag"(music, noise, sil & other), these utterances are supposed to be removed in data_prep, which we didn't.  The fix is available here https://github.com/SpeechColab/GigaSpeech/blob/860352ab639b1486e668868dea40339f4c57a8de/toolkits/kaldi/gigaspeech_data_prep.sh#L62-L66 

> Example 3)
> The entire deletion.
> Probably, in this case, the audio was too short.
> 
> ```
> Speaker sentences3741: pod1000000026_s0000036 #utts: 1
> id: (pod1000000026_s0000036-pod1000000026_s0000036)
> Scores: (#C #S #D #I) 0 1 5 0
> REF: THAT'S WHAT I SAID VERY BASIC
> HYP: ****** **** * **** **** <SOS/EOS>
> Eval: D D D D D S
> ```

this is indeed another error source in WER, I checked multiple cases and these utterances usually have really small volumes and duration, even human need to listen very carefully.  They are challenging.

At the moment, annotators are doing one more round labeling check with dev/test sets, we will update when it's ready.

**Pull request merged**